test/instrumentation/decode_metric.go

```diff
@@ -0,0 +1,238 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package main
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"sort"
+	"strconv"
+	"strings"
+
+	"k8s.io/component-base/metrics"
+)
+
+func DecodeMetricCalls(fs []*ast.CallExpr, metricsImportName string) ([]metric, []error) {
+	finder := metricDecoder{
+		metricsImportName: metricsImportName,
+	}
+	ms := make([]metric, 0, len(fs))
+	errors := []error{}
+	for _, f := range fs {
+		m, err := finder.decodeNewMetricCall(f)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		ms = append(ms, m)
+	}
+	return ms, errors
+}
+
+type metricDecoder struct {
+	metricsImportName string
+}
+
+func (c *metricDecoder) decodeNewMetricCall(fc *ast.CallExpr) (metric, error) {
+	var m metric
+	var err error
+	se, ok := fc.Fun.(*ast.SelectorExpr)
+	if !ok {
+		return m, newDecodeErrorf(fc, errNotDirectCall)
+	}
+	functionName := se.Sel.String()
+	selector, ok := se.X.(*ast.Ident)
+	if !ok {
+		return m, newDecodeErrorf(fc, errNotDirectCall)
+	}
+	if selector.String() != c.metricsImportName {
+		return m, newDecodeErrorf(fc, errNotDirectCall)
+	}
+	switch functionName {
+	case "NewCounter", "NewGauge", "NewHistogram":
```

Comment:
Maybe define these as constants at the top? It can prevent against typos since it seems like we reuse these in a few spots.